storage-s3-dynamodb/src/main/java/io/delta/storage/RetryableCloseableIterator.java

```diff
@@ -0,0 +1,159 @@
+        this.iterSupplier = Objects.requireNonNull(iterSupplier);
+        this.lastSuccessfullIndex = -1;
+        this.currentIter = this.iterSupplier.get();
+    }
+
+    /** Visible for testing. */
+    public int getLastSuccessfullIndex() {
+        return lastSuccessfullIndex;
+    }
+
+    /** Visible for testing. */
+    public int getNumRetries() {
+        return numRetries;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (currentIter != null) {
+            currentIter.close();
+        }
+    }
+
+    /**
+     * `hasNext` must be idempotent. It does not change the `lastSuccessfulIndex` variable.
+     */
+    @Override
+    public boolean hasNext() {
+        try {
+            return hasNextInternal();
+        } catch (RemoteFileChangedException ex) {
+            LOG.warn(
+                "Caught a RemoteFileChangedException in `hastNext`. NumRetries is {} / {}.\n{}",
+                numRetries + 1, MAX_RETRIES, ex.toString()
+            );
+            if (numRetries < MAX_RETRIES) {
+                numRetries++;
+                replayIterToLastSuccessfulIndex();
+                // Now, the currentImpl has been recreated and iterated to the same index
+                return hasNext();
+            } else {
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    /** Throw a checked exception so we can catch this in the caller. */
+    private boolean hasNextInternal() throws RemoteFileChangedException {
+        return currentIter.hasNext();
+    }
+
+    @Override
+    public String next() {
+        if (!hasNext()) throw new NoSuchElementException();
+
+        try {
+            final String ret = nextInternal();
+            lastSuccessfullIndex++;
+            return ret;
+        } catch (RemoteFileChangedException ex) {
+            LOG.warn(
+                "Caught a RemoteFileChangedException in `next`. NumRetries is {} / {}.\n{}",
+                numRetries + 1, MAX_RETRIES, ex.toString()
+            );
+            if (numRetries < MAX_RETRIES) {
+                numRetries++;
+                replayIterToLastSuccessfulIndex();
+                // Now, the currentImpl has been recreated and iterated to the same index
+                return next();
+            } else {
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    /** Throw a checked exception so we can catch this in the caller. */
+    private String nextInternal() throws RemoteFileChangedException {
+        return currentIter.next();
+    }
+
+    /**
+     * Called after a RemoteFileChangedException was thrown. Tries to replay the underlying
+     * iter implementation (supplied by the `implSupplier`) to the last successful index, so that
+     * the previous error open (hasNext, or next) can be retried.
+     *
+     * NOTE: This iter replay **itself** can throw a RemoteFileChangedException. Let's not deal
```

Comment:
Can we refactor the code like this to handle this case? (I wrote the code snippet without testing. May have bugs)

```
    @Override
    public String next() {
        if (!hasNext()) throw new NoSuchElementException();

        try {
            final String ret = nextInternal();
            lastSuccessfullIndex++;
            return ret;
        } catch (RemoteFileChangedException ex) {
            try {
                replayIterToLastSuccessfulIndex(ex);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            return next();
        }
    }

    private void fakeIOException() throws IOException {
        if (false) {
            throw new IOException();
        }
    }

    private void replayIterToLastSuccessfulIndex(IOException ex) throws IOException {
        LOG.warn(
                "Caught a RemoteFileChangedException in `next`. NumRetries is {} / {}.\n{}",
                numRetries + 1, MAX_RETRIES, ex);
        currentIter.close();
        while (numRetries < MAX_RETRIES) {
            numRetries++;
            LOG.info("Replaying until (inclusive) index {}", lastSuccessfullIndex);
            currentIter = iterSupplier.get();
            int currentIndex = 0;
            try {
                while (currentIndex < lastSuccessfullIndex) {
                    if (currentIter.hasNext()) {
                        // Disgard data that has been read
                        currentIter.next();
                        currentIndex++;
                    } else {
                        throw new IllegalStateException(
                                "a retried iterator doesn't have enough data (currentIndex=" +
                                        currentIndex + ", lastSuccessfullIndex=" + lastSuccessfullIndex + ")");
                    }
                }
                fakeIOException();
                return;
            } catch (IOException e) {
                if (e.getClass().getName().endsWith("org.apache.hadoop.fs.s3a.RemoteFileChangedException")) {
                    // Ignore
                } else {
                    throw e;
                }
            }
            LOG.info("Successfully replayed until (inclusive) index {}", lastSuccessfullIndex);
        }
        throw new UncheckedIOException(ex);
    }
```

Reply:
what is `fakeIOException();` for?

Reply:
ah I see